packages/babel-plugin-transform-async-generator-functions/src/for-await.ts

```diff
@@ -1,31 +1,21 @@
 import { types as t, template } from "@babel/core";
 import type { NodePath } from "@babel/traverse";
 
-const buildForAwait = template(`
-  async function wrapper() {
-    var ITERATOR_ABRUPT_COMPLETION = false;
-    var ITERATOR_HAD_ERROR_KEY = false;
-    var ITERATOR_ERROR_KEY;
+const buildForAwait = template.statements(`
+  var ITERATOR_KEY = GET_ITERATOR(OBJECT), STEP_KEY, NOT_DONE_KEY;
+  try {
+    for (;NOT_DONE_KEY = !(STEP_KEY = await ITERATOR_KEY.next()).done;NOT_DONE_KEY = false) {
+    }
+  } catch(e) {
+    STEP_KEY = null;
+    throw e;
+  } finally {
     try {
-      for (
-        var ITERATOR_KEY = GET_ITERATOR(OBJECT), STEP_KEY;
-        ITERATOR_ABRUPT_COMPLETION = !(STEP_KEY = await ITERATOR_KEY.next()).done;
-        ITERATOR_ABRUPT_COMPLETION = false
-      ) {
-      }
-    } catch (err) {
-      ITERATOR_HAD_ERROR_KEY = true;
-      ITERATOR_ERROR_KEY = err;
-    } finally {
-      try {
-        if (ITERATOR_ABRUPT_COMPLETION && ITERATOR_KEY.return != null) {
-          await ITERATOR_KEY.return();
-        }
-      } finally {
-        if (ITERATOR_HAD_ERROR_KEY) {
-          throw ITERATOR_ERROR_KEY;
-        }
+      if (NOT_DONE_KEY && ITERATOR_KEY.return) {
+        await ITERATOR_KEY.return();
       }
+    } catch (e) {
+      if (STEP_KEY) throw e;
```

Comment:
```suggestion
      if (STEP_KEY != null) throw e;
```

`.next()` might set `Boolean.prototype.done` to `true` and return `false` :)

Could you also rename `_notDone` to `_iteratorAbruptCompletion`, to reduce the diff and make it easier to follow the changes?

Reply:
nit: `.next()` must return an Object; otherwise agree explicit null-check looks clearer.

[ForIn/OfBodyEvaluation](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset)
> 6.c. If nextResult [is not an Object](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-object-type), throw a TypeError exception.

Reply:
> `.next()` might set `Boolean.prototype.done` to `true` and return `false` :)

I'm not sure if it's worth doing, it's really weird usage. :)

> Could you also rename _notDone to _iteratorAbruptCompletion, to reduce the diff and make it easier to follow the changes?

Since I completely changed the position of variable declarations and the logic of throwing exceptions, the changes may not be less, `_notDone` is slightly shorter and easier to read in the output.